Apache kafka (#1712)
knative/docs: docs/eventing/samples/kafka/channel/README.md

# Kafka CRD default channel

With the Apache Kafka CRD based channel installed, you can configure Knative to use `KafkaChannels` as default channel implementations. 

## Prerequisites

Install the `KafkaChannel` component on your configured cluster:
   ```
   curl -L "https://github.com/knative/eventing-contrib/releases/download/v0.8.2/kafka-channel.yaml" \
    | sed 's/REPLACE_WITH_CLUSTER_URL/my-cluster-kafka-bootstrap.my-kafka-namespace:9092/' \
    | kubectl apply --filename -
   ```

> > Note: The above assumes that you have Apache Kafka installed in the `my-kafka-namespace`.

Author: Ok, that assumes an installation of Kafka on `kafka` namespace,

following: https://strimzi.io/quickstarts/minikube/

HOWEVER, all a user really neeeds is a Kafka cluster that is accessible from the installtion of knative.

On k8s, with Strimzi, it's pretty darn easy, and the quickstart, it would be `kafka`,  @RichieEscarez

> > Note: The above assumes that you have Apache Kafka installed in the `my-kafka-namespace`.

Reviewer: I see, strimzi is required? I think thats where Im having trouble. We should add that to the prerequisites section and include instructions for those dependencies.

> > Note: The above assumes that you have Apache Kafka installed in the `my-kafka-namespace`.

Author: OK, let me update it.

I can also script some install options, if that's OK, @RichieEscarez

> Now we have the the `KafkaChannel` API available, we could use it like:

Reviewer 2: ```suggestion
Once the `KafkaChannel` API is available, create a new object by configuring the YAML file as follows:
```

> Once the `KafkaChannel` API is available, create a new object by configuring the YAML file as follows:

Author: @RichieEscarez  would something like

```
cat <<-EOF | kubectl apply -f -
---
apiVersion: messaging.knative.dev/v1alpha1
kind: KafkaChannel
metadata:
  name: my-kafka-channel
spec:
  numPartitions: 1
  replicationFactor: 3
EOF
```

be acceptable ?

> Once the `KafkaChannel` API is available, create a new object by configuring the YAML file as follows:

Reviewer: Yes, that works. Thanks @matzew !

```
apiVersion: messaging.knative.dev/v1alpha1
kind: KafkaChannel
metadata:
  name: my-kafka-channel
spec:
  numPartitions: 1
  replicationFactor: 3
```

Use the following tutorial to set up the `KafkaChannel` as the default channel.

Default channel configuration

> To configure the usage of the `KafkaChannel` as the _default channel_, edit the `default-ch-webhook` ConfigMap as follows:

Reviewer: ```suggestion
To configure the usage of the `KafkaChannel` CRD as the [default channel configuration](channels/default-channels.md), edit the `default-ch-webhook` ConfigMap as follows:
```

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: default-ch-webhook
  namespace: knative-eventing
data:
  # Configuration for defaulting channels that do not specify CRD implementations.
  default-ch-config: |
    clusterDefault:
      apiVersion: messaging.knative.dev/v1alpha1
      kind: KafkaChannel
```

> ## Channel CRD

Reviewer: ```suggestion
## Creating an Apache Kafka channel
```

> Use the `channels.messaging.knative.dev` CRD to create a new channel:

Reviewer: ```suggestion
Now that `KafkaChannel` is set as the default channel configuration, you can use the `channels.messaging.knative.dev` CRD to create a new Apache Kafka channel:
```

```
apiVersion: messaging.knative.dev/v1alpha1
kind: Channel
metadata:
  name: testchannel-one
```

NOTE: This does not allow you to set specific details of the actual channel implementation, such as partitions on the Kafka topic. To set specific details use the `KafkaChannel` API instead.

Now, you can check in Kafka if there is a `testchannel` topic. With Strimzi this can be done like:

```
kubectl -n kafka exec -it my-cluster-kafka-0 -- bin/kafka-topics.sh --zookeeper localhost:2181 --list
```

The result is:

```
...
knative-messaging-kafka.default.testchannel-one
...
```

The Apache Kafka topic that was created by the Channel implementatio is prefixed with `knative-messaging-kafka`, to indicate this is an Apache Kafka channel from Knative. It than contains the name of the namespace (`default` in our case) followed by the actual name of the channel.

> ## Apache Kafka as channel for the Knative broker

Reviewer: ```suggestion
## Configuring the Knative broker for Apache Kafka channels
```

To setup a broker that will use the new default Kafka channels, you must inject a new _default_ broker, using the command:

```
kubectl label namespace default knative-eventing-injection=enabled
```

This will give you two pods, like:

```
default-broker-filter-64658fc79f-nf596                            1/1     Running     0          15m
default-broker-ingress-ff79755b6-vj9jt                            1/1     Running     0          15m

```
Inside the Apache Kafka cluster you might see two new topics, such as:

```
...
knative-messaging-kafka.default.default-kn2-ingress
knative-messaging-kafka.default.default-kn2-trigger
...
```

> ## Demo 

Reviewer: this reads more like the necessary steps for using the broker?

```suggestion
## Creating a service and trigger to use the Apache Kafka broker
```

To use the Apache Kafka based broker, let's take a look at a simple demo. Use the`ApiServerSource` to publish events to the broker as well as the `Trigger` API, which then routes events to a Knative `Service`.

> `ApiServerSource` to publish events to the broker as well as the `Trigger` API which than routes events

Reviewer: remove duplicates
```suggestion
```

> to a Knative `Service`.

Reviewer: ```suggestion
```

1. Install `ksvc`, using the command:
    ```
    kubectl apply -f 000-ksvc.yaml
    ```
2. Install a source that publishes to the default broker
    ```
    kubectl apply -f 020-k8s-events.yaml
    ```

> 3. Have a trigger that routes the events to the `ksvc`:

Reviewer: ```suggestion
3. Create a trigger that routes the events to the `ksvc`:
```

    ```
    kubectl apply -f 030-trigger.yaml
    ```

> 

Reviewer: ```suggestion

## Verifying your Apache Kafka channel and broker

 Now that your Eventing cluster is configured for Apache Kafka, you can verify
 your configuration with the following options.
 
```

You must also have the following tools installed:
- `curl`
- `sed`

## Creating a `KafkaChannel` channel CRD

Install the `KafkaChannel` sub-component on your Knative Eventing cluster:
   ```
   curl -L "https://github.com/knative/eventing-contrib/releases/download/v0.8.2/kafka-channel.yaml" \
    | sed 's/REPLACE_WITH_CLUSTER_URL/my-cluster-kafka-bootstrap.kafka:9092/' \
    | kubectl apply --filename -
   ```

> Note: The above assumes that you have Apache Kafka installed in the `kafka`, as discussed [here](../README.md)!

Once the `KafkaChannel` API is available, create a new object by configuring the YAML file as follows:

```
cat <<-EOF | kubectl apply -f -
---
apiVersion: messaging.knative.dev/v1alpha1
kind: KafkaChannel
metadata:
  name: my-kafka-channel
spec:
  numPartitions: 1
  replicationFactor: 3
EOF
```

You can now set the `KafkaChannel` CRD as the default channel configuration.

## Specifying the default channel configuration

To configure the usage of the `KafkaChannel` CRD as the [default channel configuration](channels/default-channels.md), edit the `default-ch-webhook` ConfigMap as follows:

> ```

Reviewer: Can this be applied in the same way? 
ie.
```
cat <<-EOF | kubectl apply -f -
---
apiVersion: v1 ...
 ...
EOF
```

Reviewer 3: Yes

# Apache Kafka CRD default channel

You can install and configure the Apache Kafka CRD (`KafkaChannel`) as the default channel configuration in Knative Eventing. 

## Prerequisites

> You must ensure that you meet the [prerequisites listed in the Apache Kafka overview](../README.md).

Author: @RichieEscarez  the referenced doc there does actually already link to install instructions for Strimzi, on k8s

Reviewer: Thanks! Ive added comments in that file.

---
title: "Apache Kafka CRD Example"
linkTitle: "CRD - default channel"
weight: 30
type: "docs"
---

You can install and configure the Apache Kafka CRD (`KafkaChannel`) as the default channel configuration in Knative Eventing. 

## Prerequisites

You must ensure that you meet the [prerequisites listed in the Apache Kafka overview](../README.md).

You must also have the following tools installed:
- `curl`
- `sed`

## Creating a `KafkaChannel` channel CRD

Install the `KafkaChannel` sub-component on your Knative Eventing cluster:
   ```

>    curl -L "https://github.com/knative/eventing-contrib/releases/download/v0.8.2/kafka-channel.yaml" \

Reviewer 4: Should this be v0.9.0?

    | sed 's/REPLACE_WITH_CLUSTER_URL/my-cluster-kafka-bootstrap.kafka:9092/' \
    | kubectl apply --filename -
   ```

> Note: The above assumes that you have Apache Kafka installed in the `kafka`, as discussed [here](../README.md)!

Once the `KafkaChannel` API is available, create a new object by configuring the YAML file as follows:

```
cat <<-EOF | kubectl apply -f -
---
apiVersion: messaging.knative.dev/v1alpha1
kind: KafkaChannel
metadata:
  name: my-kafka-channel
spec:
  numPartitions: 1
  replicationFactor: 3
EOF
```

You can now set the `KafkaChannel` CRD as the default channel configuration.

## Specifying the default channel configuration

To configure the usage of the `KafkaChannel` CRD as the [default channel configuration](channels/default-channels.md), edit the `default-ch-webhook` ConfigMap as follows:

```
cat <<-EOF | kubectl apply -f -
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: default-ch-webhook
  namespace: knative-eventing
data:
  # Configuration for defaulting channels that do not specify CRD implementations.
  default-ch-config: |
    clusterDefault:
      apiVersion: messaging.knative.dev/v1alpha1
      kind: KafkaChannel
EOF
```

> ## Creating an Apache Kafka channel

Reviewer 4: Optional nit: Maybe this title should be specific to the default case, e.g.

```
## Creating an Apache Kafka channel using the default channel configuration
```
```

---
apiVersion: messaging.knative.dev/v1alpha1
kind: Channel
metadata:
  name: testchannel-one
EOF
```

Check Kafka for a `testchannel` topic. With Strimzi this can be done by using the command:

```
kubectl -n kafka exec -it my-cluster-kafka-0 -- bin/kafka-topics.sh --zookeeper localhost:2181 --list
```

The result is:

```
...
knative-messaging-kafka.default.testchannel-one
...
```

The Apache Kafka topic that is created by the channel implementation is prefixed with `knative-messaging-kafka`. This indicates it is an Apache Kafka channel from Knative. It contains the name of the namespace, `default` in this example, followed by the actual name of the channel.

## Configuring the Knative broker for Apache Kafka channels

To setup a broker that will use the new default Kafka channels, you must inject a new _default_ broker, using the command:

```
kubectl label namespace default knative-eventing-injection=enabled
```

This will give you two pods, such as:

```
default-broker-filter-64658fc79f-nf596                            1/1     Running     0          15m
default-broker-ingress-ff79755b6-vj9jt                            1/1     Running     0          15m

```

> Inside the Apache Kafka cluster you might see two new topics, such as:

Reviewer 4: Optional nit: This is a _should_ right? you should see two new topics. If you don't, that indicates a problem.

```
...
knative-messaging-kafka.default.default-kn2-ingress
knative-messaging-kafka.default.default-kn2-trigger
...
```

## Creating a service and trigger to use the Apache Kafka broker

To use the Apache Kafka based broker, let's take a look at a simple demo. Use the`ApiServerSource` to publish events to the broker as well as the `Trigger` API, which then routes events to a Knative `Service`.

> 1. Install `ksvc`, using the command:

Reviewer 4: Optional nit: This might be easier to understand if it used the same term as above (Knative `Service`) instead of the new term `ksvc`.

    ```
    kubectl apply -f 000-ksvc.yaml
    ```
2. Install a source that publishes to the default broker
    ```
    kubectl apply --filename 020-k8s-events.yaml
    ```

> 3. Create a trigger that routes the events to the `ksvc`:

Reviewer 4: Optional nit: Same as above re `ksvc`

    ```
    kubectl apply -f 030-trigger.yaml
    ```

## Verifying your Apache Kafka channel and broker

 Now that your Eventing cluster is configured for Apache Kafka, you can verify
 your configuration with the following options.
 
### Receive events via Knative

> Now you can see the events in the log of the `ksvc`, using the command:

Reviewer 4: Optional nit: No need for the comma here
